File: lib/models/chess_timer.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../global.dart';

/// Represents the state of a chess timer
enum TimerState {
  /// Timer is not running and has not been started
  ready,

  /// Timer is currently running
  running,

  /// Timer is paused
  paused,

  /// Timer has expired (player lost on time)
  expired,

  /// Timer is stopped (game ended)
  stopped
}

/// A chess timer model for blitz games
class ChessTimer extends ChangeNotifier {
  /// Initial time in seconds (3 minutes = 180 seconds)
  final int initialTime;

  /// Increment in seconds (2 seconds per move)
  final int increment;

  /// Current time remaining in seconds
  int _timeRemaining;

  /// Accumulated fractional seconds for more accurate timing
  double _fractionalSeconds = 0.0;

  /// Current state of the timer
  TimerState _state = TimerState.ready;

  /// Timer for counting down
  Timer? _timer;

  /// Last time the timer was updated (for accurate timing)
  DateTime? _lastUpdateTime;

  /// Constructor
  ChessTimer({
    this.initialTime = 180, // 3 minutes
    this.increment = 2,     // 2 seconds per move
  }) : _timeRemaining = initialTime;

  /// Get the current time remaining in seconds
  int get timeRemaining => _timeRemaining;

  /// Get the current state of the timer
  TimerState get state => _state;

  /// Check if the timer is running
  bool get isRunning => _state == TimerState.running;

  /// Check if the timer has expired
  bool get isExpired => _state == TimerState.expired;

  /// Format the time remaining as mm:ss
  String get formattedTime {
    final minutes = (_timeRemaining ~/ 60).toString().padLeft(2, '0');
    final seconds = (_timeRemaining % 60).toString().padLeft(2, '0');
    return '$minutes:$seconds';
  }

  /// Start the timer
  void start() {
    if (_state == TimerState.running) return;

    _state = TimerState.running;
    _lastUpdateTime = DateTime.now();

    _timer?.cancel();
    // Reduced frequency from 100ms to 1000ms for better performance
    _timer = Timer.periodic(const Duration(milliseconds: 1000), _updateTimer);

    notifyListeners();
    logger.info('Timer started: $_timeRemaining seconds');
  }

  /// Pause the timer
  void pause() {
    if (_state != TimerState.running) return;

    _state = TimerState.paused;
    _timer?.cancel();
    _timer = null;

    notifyListeners();
    logger.info('Timer paused: $_timeRemaining seconds');
  }

  /// Resume the timer
  void resume() {
    if (_state != TimerState.paused) return;

    start();
  }

  /// Stop the timer (game ended)
  void stop() {
    _state = TimerState.stopped;
    _timer?.cancel();
    _timer = null;

    notifyListeners();
    logger.info('Timer stopped: $_timeRemaining seconds');
  }

  /// Reset the timer to the initial time
  void reset() {
    _state = TimerState.ready;
    _timeRemaining = initialTime;
    _fractionalSeconds = 0.0;
    _timer?.cancel();
    _timer = null;

    notifyListeners();
    logger.info('Timer reset: $_timeRemaining seconds');
  }

  /// Add increment to the timer (after a move)
  void addIncrement() {
    if (_state == TimerState.expired || _state == TimerState.stopped) return;

    _timeRemaining += increment;
    notifyListeners();
    logger.info('Increment added: $_timeRemaining seconds');
  }

  /// Set the time remaining (for synchronization)
  void setTimeRemaining(int seconds) {
    _timeRemaining = seconds;

    if (_timeRemaining <= 0) {
      _timeRemaining = 0;
      _state = TimerState.expired;
      _timer?.cancel();
      _timer = null;
    }

    notifyListeners();
    logger.info('Time set: $_timeRemaining seconds');
  }

  /// Update the timer (called by the timer)
  void _updateTimer(Timer timer) {
    if (_state != TimerState.running) return;

    final now = DateTime.now();
    final elapsed = now.difference(_lastUpdateTime!).inMilliseconds;
    _lastUpdateTime = now;

    // Convert elapsed milliseconds to seconds
    final elapsedSeconds = elapsed / 1000.0;

    // Add to fractional seconds accumulator
    _fractionalSeconds += elapsedSeconds;

    // If we've accumulated a full second or more, subtract from time remaining
    if (_fractionalSeconds >= 1.0) {
      final secondsToSubtract = _fractionalSeconds.floor();
      final previousTime = _timeRemaining;
      _timeRemaining -= secondsToSubtract;
      _fractionalSeconds -= secondsToSubtract;

      // Only notify listeners if the displayed time actually changed
      final timeChanged = previousTime != _timeRemaining;

      // Ensure time doesn't go below zero
      if (_timeRemaining <= 0) {
        _timeRemaining = 0;
        _fractionalSeconds = 0.0;
        _state = TimerState.expired;
        _timer?.cancel();
        _timer = null;
        logger.info('Timer expired');
        notifyListeners(); // Always notify on expiration
      } else if (timeChanged) {
        notifyListeners(); // Only notify if time display changed
      }
    }
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}

File: src/models/ChessTimer.ts
/**
 * Enum representing the possible states of a chess timer
 */
export enum TimerState {
  READY = 'ready',
  RUNNING = 'running',
  PAUSED = 'paused',
  EXPIRED = 'expired',
  STOPPED = 'stopped',
}

/**
 * Class representing a chess timer for a single player
 */
export class ChessTimer {
  private initialTimeSeconds: number;
  private incrementSeconds: number;
  private timeRemainingSeconds: number;
  private timerState: TimerState;
  private timerInterval: NodeJS.Timeout | null;
  private lastUpdateTime: Date | null;
  private listeners: Array<() => void>;

  /**
   * Constructor for ChessTimer
   * @param initialTimeSeconds Initial time in seconds (default: 180 seconds = 3 minutes)
   * @param incrementSeconds Time increment in seconds after each move (default: 2 seconds)
   */
  constructor(initialTimeSeconds: number = 180, incrementSeconds: number = 2) {
    this.initialTimeSeconds = initialTimeSeconds;
    this.incrementSeconds = incrementSeconds;
    this.timeRemainingSeconds = initialTimeSeconds;
    this.timerState = TimerState.READY;
    this.timerInterval = null;
    this.lastUpdateTime = null;
    this.listeners = [];
  }

  /**
   * Add a listener to be notified of timer changes
   * @param listener Function to call when timer changes
   */
  addListener(listener: () => void): void {
    this.listeners.push(listener);
  }

  /**
   * Remove a listener
   * @param listener Function to remove
   */
  removeListener(listener: () => void): void {
    this.listeners = this.listeners.filter(l => l !== listener);
  }

  /**
   * Notify all listeners of a change
   */
  private notifyListeners(): void {
    this.listeners.forEach(listener => listener());
  }

  /**
   * Start the timer
   */
  start(): void {
    if (this.timerState === TimerState.RUNNING) {
      return;
    }

    this.timerState = TimerState.RUNNING;
    this.lastUpdateTime = new Date();

    // Clear any existing interval
    if (this.timerInterval) {
      clearInterval(this.timerInterval);
    }

    // Update the timer every 100ms for smooth display
    this.timerInterval = setInterval(() => {
      if (this.timerState !== TimerState.RUNNING) {
        return;
      }

      const now = new Date();
      if (this.lastUpdateTime) {
        const elapsedSeconds = (now.getTime() - this.lastUpdateTime.getTime()) / 1000;
        this.timeRemainingSeconds -= elapsedSeconds;

        // Check if timer has expired
        if (this.timeRemainingSeconds <= 0) {
          this.timeRemainingSeconds = 0;
          this.timerState = TimerState.EXPIRED;
          this.stop();
        }
      }

      this.lastUpdateTime = now;
      this.notifyListeners();
    }, 100);

    this.notifyListeners();
  }

  /**
   * Pause the timer
   */
  pause(): void {
    if (this.timerState !== TimerState.RUNNING) {
      return;
    }

    this.timerState = TimerState.PAUSED;
    this.lastUpdateTime = null;

    if (this.timerInterval) {
      clearInterval(this.timerInterval);
      this.timerInterval = null;
    }

    this.notifyListeners();
  }

  /**
   * Stop the timer
   */
  stop(): void {
    this.timerState = TimerState.STOPPED;
    this.lastUpdateTime = null;

    if (this.timerInterval) {
      clearInterval(this.timerInterval);
      this.timerInterval = null;
    }

    this.notifyListeners();
  }

  /**
   * Reset the timer to initial time
   */
  reset(): void {
    this.timeRemainingSeconds = this.initialTimeSeconds;
    this.timerState = TimerState.READY;
    this.lastUpdateTime = null;

    if (this.timerInterval) {
      clearInterval(this.timerInterval);
      this.timerInterval = null;
    }

    this.notifyListeners();
  }

  /**
   * Add increment time after a move
   */
  addIncrement(): void {
    this.timeRemainingSeconds += this.incrementSeconds;
    this.notifyListeners();
  }

  /**
   * Set the time remaining
   * @param seconds Time in seconds
   */
  setTimeRemaining(seconds: number): void {
    this.timeRemainingSeconds = seconds;
    this.notifyListeners();
  }

  /**
   * Get the time remaining in seconds
   */
  getTimeRemaining(): number {
    return this.timeRemainingSeconds;
  }

  /**
   * Get the timer state
   */
  getState(): TimerState {
    return this.timerState;
  }

  /**
   * Check if the timer has expired
   */
  isExpired(): boolean {
    return this.timerState === TimerState.EXPIRED;
  }

  /**
   * Get formatted time string (mm:ss)
   */
  getFormattedTime(): string {
    const minutes = Math.floor(this.timeRemainingSeconds / 60);
    const seconds = Math.floor(this.timeRemainingSeconds % 60);
    return `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
  }

  /**
   * Clean up resources
   */
  dispose(): void {
    if (this.timerInterval) {
      clearInterval(this.timerInterval);
      this.timerInterval = null;
    }
    this.listeners = [];
  }
}
